source/extensions/filters/http/cache/http_cache_utils.cc

```diff
@@ -0,0 +1,175 @@
+#include "extensions/filters/http/cache/http_cache_utils.h"
+
+#include <array>
+#include <string>
+
+#include "absl/algorithm/container.h"
+#include "absl/strings/ascii.h"
+#include "absl/strings/numbers.h"
+#include "absl/strings/strip.h"
+
+using absl::ascii_isalnum;
+using absl::ascii_isdigit;
+using absl::Duration;
+using absl::InfiniteDuration;
+using absl::InfinitePast;
+using absl::ParseTime;
+using absl::Seconds;
+using absl::SimpleAtoi;
+using absl::string_view;
+using absl::Time;
+using absl::ZeroDuration;
+using Envoy::Http::HeaderEntry;
+using std::array;
+using std::string;
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace Cache {
+namespace Internal {
+
+namespace {
+
+bool tchar(char c) {
+  switch (c) {
+  case '!':
+  case '#':
+  case '$':
+  case '%':
+  case '&':
+  case '*':
+  case '+':
+  case '-':
+  case '.':
+  case '^':
+  case '_':
+  case '`':
+  case '|':
+  case '~':
+    return true;
+  }
+  return ascii_isalnum(c);
+}
+
+bool eatToken(string_view& s) {
+  const string_view::iterator token_end = c_find_if_not(s, &tchar);
+  if (token_end == s.begin()) {
+    return false;
+  }
+  s.remove_prefix(token_end - s.begin());
+  return true;
+}
+
+void eatDirectiveArgument(string_view& s) {
+  if (s.empty()) {
+    return;
+  }
+  if (s.front() == '"') {
+    // TODO(toddmgreer) handle \-escaped quotes
+    const size_t closing_quote = s.find('"', 1);
+    s.remove_prefix(closing_quote);
+  } else {
+    eatToken(s);
+  }
+}
+} // namespace
+
+// If s is nonnull and begins with decimal digits, return Eat leading digits in
+// *s, if any
+Duration eatLeadingDuration(string_view* s) {
+  const string_view::iterator digits_end = c_find_if_not(*s, &ascii_isdigit);
+  const size_t digits_length = digits_end - s->begin();
+  if (digits_length == 0) {
+    return ZeroDuration();
+  }
+  const string_view digits(s->begin(), digits_length);
+  s->remove_prefix(digits_length);
+  uint64_t num;
+  return SimpleAtoi(digits, &num) ? Seconds(num) : InfiniteDuration();
+}
+
+Duration effectiveMaxAge(string_view cache_control) {
+  // The grammar for This Cache-Control header value should be:
```

Comment:
can you give a link to where this grammar is specified, e.g. http spec or something more digestible, which I've also seen on the web?

Reply:
Done.